library/core/src/option.rs

```diff
@@ -70,10 +70,23 @@
 //! }
 //! ```
 //!
-//! This usage of [`Option`] to create safe nullable pointers is so
-//! common that Rust does special optimizations to make the
-//! representation of [`Option`]`<`[`Box<T>`]`>` a single pointer. Optional pointers
-//! in Rust are stored as efficiently as any other pointer type.
+//! # Representation
+//!
+//! Rust guarantees to optimize the following types `<T>` such that an
+//! [`Option<T>`] has the same size as `T`:
+//!
+//! * [`Box<T>`]
+//! * `&T`
+//! * `&mut T`
+//! * `extern "C" fn`
```

Comment:
```suggestion
//! * `fn`, `extern "C" fn`
```
Seems odd to exclude rust function pointers. I didn't realize this before.

Reply:
Good point, I had overlooked that too!